walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -22,59 +22,187 @@
 import java.util.List;
 
 public final class FindMeetingQuery {
-  
-  private Collection<TimeRange> possibleTimes ;
 
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    //throw new UnsupportedOperationException("TODO: Implement this method.");
-    // If no attndees
-    if ((request.getAttendees() == null ) && (events == null) ){
+
+    // Convert collection of events into an arraylist.
+    ArrayList<Event> eventList = new ArrayList<>(events);
+    
+    // Initialise result variable. 
+    Collection<TimeRange> possibleTimes = new ArrayList<>();
+    // Result considering only mandatory attendees.
+    Collection<TimeRange> timesMandatory = new ArrayList<>();
+
+    // Initialise variables for no events and no attendees.
+    Collection<Event> NO_EVENTS = Collections.emptySet();
+    Collection<String> NO_ATTENDEES = Collections.emptySet();
+
+    
+    // Check if there are optional attendee, who is busy and can't attend.
+    for (int i=0; i< eventList.size(); i++) {
+      if ( ( (eventList.get(i).getAttendees() .containsAll( request.getOptionalAttendees() )) == true ) && (eventList.get(i).getWhen() == TimeRange.WHOLE_DAY) ) {
+          eventList.remove(eventList.get(i)); 
```

Comment:
Removing things from a list you're iterating through is bad practice and prone to errors, for instance if you remove list[0], then list[1] moves to list[0] and you don't do this check on it.
There are two ways you can achieve the same thing here, either loop through the list and add the elements you want to keep to a new list, or even easier just use ArrayList.removeif() and it'll handle the removing in a safe way for you.